cachi2/core/package_managers/gomod.py

```diff
@@ -537,29 +596,63 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[ParsedPackage]:
 
     _validate_local_replacements(all_modules, app_dir)
 
-    return main_module, all_modules, all_packages
+    return ResolveResult(main_module, all_modules, all_packages, modules_in_go_sum)
+
+
+def _parse_go_sum(module_dir: RootedPath) -> frozenset[ModuleID]:
+    """Return the set of modules present in the go.sum file in the specified directory.
+
+    A module is considered present if the checksum for its .zip file is present. The go.mod file
+    checksums are not relevant for our purposes.
```

Comment:
> go.mod file checksums are not relevant

Not entirely true, but that's a long story.

Go uses the .mod files to do dependency resolution. If we're not sure that all the .mod files had their checksums verified, we can't be sure that the resolved dependencies are the same as they were last time.

But Go also looks at the .mod files of modules that are not needed, while we aim to report only those that are. So even if we took .mod files into account, we wouldn't be able to report whether dependency resolution was 100% checksum-verified.

Reply:
And for what it's worth, the .zip file for a module does include its .mod file, so attempts to change the .mod file in the source of a dependency should be caught (but again, only for the modules that are actually needed for the build)

Reply:
> Go uses the .mod files to do dependency resolution.

Is this why there are some modules in the go.sum file that only have hashes for the go.mod file (and not for the zip file)? [e.g.](https://github.com/redhat-appstudio/build-service/blob/main/go.sum#L2)

Reply:
I think so. But I've even seen the opposite case - a module that only has the .zip checksum (and I don't really know what that means - maybe old modules that don't have go.mod)